### PR TITLE
[generator] Add C#8 nullability into generated code

### PR DESCRIPTION
**Pull request by spouliot**
Goals
* Reflect Apple nullability annotations in our bindings using C#8
* No warnings when building bindings

Non-Goals
* Update (add or fix) `[NullAllowed]` to match Apple headers (next phase)
* Make the generator or internal code fully nullable aware (`nowarn` is used)

Notes
* Apple's own annotations are not 100% accurate :(
* Where known issue exists we have _fixed_ our attributes to match reality :)
* We also do additional null-checks internally that might seems not required (better safe than sorry).

TODO
* ~some (generated?) .csproj needs to allow the latest C# language version~
* ~use `-warnaserror` so we don't mess up nullability in future binding updates~

### CODE REVIEW

**Comment by spouliot** [comment]
build

**Comment by spouliot** [comment]
Test results gives 404 :(

**Comment by spouliot** [comment]
build

File: src/generator.cs [hunk 1]
```diff
@@ -5084,6 +5121,8 @@ void GenerateAsyncMethod (MemberInformation original_minfo, AsyncMethodKind asyn
 			if (tuple)
 				ttype = "Tuple<bool,NSError>";
 		}
+		if (ttype == "global::FileProvider.INSFileProviderItem")
```
Comment (chamons) [inline src/generator.cs:5124]: Debug spam?

File: src/generator.cs [hunk 2]
```diff
@@ -7006,10 +7053,11 @@ public void Generate (Type type)
 							} else
 								eaname = "<NOTREACHED>";
 							
-							if (bta.Singleton || mi.GetParameters ().Length == 1)
-								print ("EventHandler handler = {0};", PascalCase (miname));
-							else
-								print ("EventHandler<{0}> handler = {1};", GetEventArgName (mi), miname);
+							print ("var handler = {0};", PascalCase (miname));
+							// if (bta.Singleton || mi.GetParameters ().Length == 1)
```
Comment (chamons) [inline src/generator.cs:7057]: Commented out?

Comment (spouliot) [reply src/generator.cs:7057]: yup, forgot to remove it. It's now using `var`, which is simpler than detecting the nullability requirement.

File: src/generator.cs [hunk 3]
```diff
@@ -4342,9 +4365,13 @@ public void GenerateMethodBody (MemberInformation minfo, MethodInfo mi, string s
 				print ("IntPtr ret;");
 			} else if (minfo.is_bindAs) {
 				var bindAsAttrib = GetBindAsAttribute (minfo.mi);
+				// tricky, e.g. when an nullable `NSNumber[]` is bound as a `float[]`, since FormatType and bindAsAttrib have not clue about the original nullability 
 				print ("{0} ret;", FormatType (bindAsAttrib.Type.DeclaringType, GetCorrectGenericType (bindAsAttrib.Type)));
-			} else
+//				print ("#pragma warning disable 8600");
```
Comment (mandel-macaque) [inline src/generator.cs:4370]: Comment left on purpose?

File: src/generator.cs [hunk 4]
```diff
@@ -4342,9 +4365,13 @@ public void GenerateMethodBody (MemberInformation minfo, MethodInfo mi, string s
...
+			} else {
 				print ("{0} ret;", FormatType (mi.DeclaringType, GetCorrectGenericType (mi.ReturnType))); //  = new {0} ();"
+//				print ("#pragma warning disable 8600");
```
Comment (mandel-macaque) [inline src/generator.cs:4373]: Same

File: src/generator.cs [hunk 5]
```diff
@@ -4473,14 +4500,16 @@ public void GenerateMethodBody (MemberInformation minfo, MethodInfo mi, string s
...
 					// FIXME: https://trello.com/c/1ukS9TbL/43-introduce-common-object-type-for-all-unmanaged-types-which-will-correctly-implement-idisposable-and-inativeobject
 					// We should consider using return INativeObject<T> (ptr, bool); here at some point
-					print ("return ret == IntPtr.Zero ? null : new global::{0} (ret);", mi.ReturnType.FullName);
+					print ("return ret == IntPtr.Zero ? null! : new global::{0} (ret);", mi.ReturnType.FullName);
 				}
 			} else {
-				print ("return ret;");
+				// we can't be 100% confident that the ObjC API annotations are correct so we always null check inside generated code
+				print ("return ret!;");
+//				print ("#pragma warning restore 8600");
```
Comment (mandel-macaque) [inline src/generator.cs:4512]: Same

File: src/generator.cs [hunk 6]
```diff
@@ -6644,11 +6688,13 @@ public void Generate (Type type)
...
 						fieldTypeName = FormatType (field_pi.DeclaringType, field_pi.PropertyType);
 
+					bool nullable = false;
 					// Value types we dont cache for now, to avoid Nullable<T>
 					if (!field_pi.PropertyType.IsValueType || smartEnumTypeName != null) {
 						print_generated_code ();
 						PrintPreserveAttribute (field_pi);
-						print ("static {0} _{1};", fieldTypeName, field_pi.Name);
+						print ("static {0}? _{1};", fieldTypeName, field_pi.Name);
+						//nullable = true;
```
Comment (mandel-macaque) [inline src/generator.cs:6697]: Same.

File: src/generator.csproj [hunk 7]
```diff
@@ -11,6 +11,7 @@
     <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
     <BuildDir Condition="'$(BUILD_DIR)' != ''">$(BUILD_DIR)\</BuildDir>
     <BuildDir Condition="'$(BUILD_DIR)' == ''">build\</BuildDir>
+    <LangVersion>latest</LangVersion>
```
Comment (mandel-macaque) [inline src/generator.csproj:14]: So, I hope this does not hit us. I have seen issues in CI were latests was not 8.0 :/

File: tests/BundledResources/BundledResources.csproj [hunk 8]
```diff
@@ -32,6 +32,33 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release-bitcode|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64_32|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
```
Comment (mandel-macaque) [inline tests/BundledResources/BundledResources.csproj:61]: Same regarding CI, I also wonder, do we have to add the LangVersion in each config? I think we can set it at the project level, but I might be wrong.

File: tests/EmbeddedResources/EmbeddedResources.csproj [hunk 9]
```diff
@@ -34,6 +34,33 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release-bitcode|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64_32|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
```
Comment (mandel-macaque) [inline tests/EmbeddedResources/EmbeddedResources.csproj:63]: Same.

File: tests/bindings-framework-test/bindings-framework-test.csproj [hunk 10]
```diff
@@ -53,6 +53,33 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release-bitcode|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64_32|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
```
Comment (mandel-macaque) [inline tests/bindings-framework-test/bindings-framework-test.csproj:82]: Same.

File: tests/bindings-test/bindings-test.csproj [hunk 11]
```diff
@@ -59,6 +59,33 @@
...
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release-bitcode|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
+  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release64_32|AnyCPU' ">
+    <LangVersion>latest</LangVersion>
+  </PropertyGroup>
```
Comment (mandel-macaque) [inline tests/bindings-test/bindings-test.csproj:88]: Same.

File: tests/cecil-tests/cecil-tests.csproj [hunk 12]
```diff
@@ -20,14 +20,14 @@
...
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>latest</LangVersion>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <Optimize>true</Optimize>
     <OutputPath>bin\Release</OutputPath>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>latest</LangVersion>
```
Comment (mandel-macaque) [inline tests/cecil-tests/cecil-tests.csproj:30]: This is what I was talking about, I think it was due to d16-6.

File: src/generator.cs (hunk 3 above)
Comment (spouliot) [reply src/generator.cs:4370]: no, some more clean up is coming :)

File: tests/BundledResources/BundledResources.csproj (hunk 8 above)
Comment (spouliot) [reply tests/BundledResources/BundledResources.csproj:61]: Maybe... it was done thru IDE

File: src/generator.csproj (hunk 7 above)
Comment (mandel-macaque) [reply src/generator.csproj:14]: I think  is when we backported xcode11.4 to d16-6, but we are not backporting this, so I don't think it will hit us this time.

File: tests/framework-test/framework-test.csproj [hunk 13]
```diff
@@ -42,6 +43,9 @@
     <MtouchFloat32>true</MtouchFloat32>
     <CodesignKey>iPhone Developer</CodesignKey>
     <MtouchUseLlvm>True</MtouchUseLlvm>
+    <DeviceSpecificBuild></DeviceSpecificBuild>
+    <MtouchNoSymbolStrip></MtouchNoSymbolStrip>
```
[the remark this answers is not in the document]
Comment (spouliot) [reply tests/framework-test/framework-test.csproj:47]: changes came from IDE :(

File: tests/linker/mac/link all/link all-mac.csproj [hunk 14]
```diff
@@ -33,6 +34,8 @@
     <AOTMode>None</AOTMode>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <MonoBundlingExtraArgs>--optimize=all,-remove-uithread-checks</MonoBundlingExtraArgs>
+    <DelaySign>false</DelaySign>
+    <CheckForOverflowUnderflow>false</CheckForOverflowUnderflow>
```
Comment (rolfbjarne) [inline tests/linker/mac/link all/link all-mac.csproj:38]: Unrelated changes:

```suggestion

```

File: tests/monotouch-test/monotouch-test.csproj [hunk 15]
```diff
@@ -61,6 +63,8 @@
     <IpaPackageName>
     </IpaPackageName>
     <MtouchNoSymbolStrip>true</MtouchNoSymbolStrip>
+    <MtouchProfiling></MtouchProfiling>
+    <MtouchVerbosity></MtouchVerbosity>
```
Comment (rolfbjarne) [inline tests/monotouch-test/monotouch-test.csproj:67]: Unrelated changes:

```suggestion
```

File: tests/monotouch-test/monotouch-test.csproj [hunk 16]
```diff
@@ -53,6 +54,7 @@
     <DefineConstants>DEBUG;DEVICE;$(DefineConstants)</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>0</WarningLevel>
+    <DocumentationFile></DocumentationFile>
```
Comment (rolfbjarne) [inline tests/monotouch-test/monotouch-test.csproj:57]: Unrelated change:

```suggestion
```

File: tests/interdependent-binding-projects/interdependent-binding-projects.csproj [hunk 17]
```diff
@@ -29,6 +29,7 @@
     <MtouchArch>i386, x86_64</MtouchArch>
     <CodesignKey>iPhone Developer</CodesignKey>
     <MtouchLink>None</MtouchLink>
+    <LangVersion>latest</LangVersion>
```
Comment (rolfbjarne) [inline tests/interdependent-binding-projects/interdependent-binding-projects.csproj:32]: Move to unconditional PropertyGroup (same for rest of file).

File: tests/interdependent-binding-projects/interdependent-binding-projects.csproj [hunk 18]
```diff
@@ -124,6 +131,9 @@
     <MtouchArch>ARM64</MtouchArch>
     <AllowUnsafeBlocks>True</AllowUnsafeBlocks>
     <MtouchUseLlvm>true</MtouchUseLlvm>
+    <IpaPackageName></IpaPackageName>
+    <NoWarn></NoWarn>
```
Comment (rolfbjarne) [inline tests/interdependent-binding-projects/interdependent-binding-projects.csproj:135]: Unrelated change:

```suggestion
```

File: tests/cecil-tests/cecil-tests.csproj [hunk 19]
```diff
@@ -20,14 +20,14 @@
     <DefineConstants>DEBUG;</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <LangVersion>8.0</LangVersion>
+    <LangVersion>latest</LangVersion>
```
Comment (rolfbjarne) [inline tests/cecil-tests/cecil-tests.csproj:23]: Move to unconditional PropertyGroup.

**Comment by spouliot** [comment]
Not sure what's up with the bot. Logs are not very helpful :(

API diff (for PR only) https://xamjenkinsartifact.blob.core.windows.net/xamarin-macios-pr-builder/13318/jenkins-results/apicomparison/api-diff.html

Generator diff https://xamjenkinsartifact.blob.core.windows.net/xamarin-macios-pr-builder/13318/jenkins-results/generator-diff/index.html

**Comment by spouliot** [comment]
Got some results at last...

It seems some projects hits a condition where the same method is enqueued over and over again (and that should not happen). Since it's a queue (and not recursive) the process can go on forever (or until the bot/process times out)

```
...
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
  Enqueue System.Void System.Runtime.CompilerServices.NullableAttribute::.ctor(System.Byte) (TaskId:105)
...
```

**Comment by spouliot** [comment]
@monojenkins backport to d16-7
